### ISSUE TITLE
Native AOT with .NET 8

### ISSUE THREAD

**Issue by SbiCA** [rebus-org/Rebus#1095]
Hi @mookid8000

Did you already try out using rebus with native AOT in .NET 8?

Do you already have an idea what's going to break? I guess the Router and Handler discovery might be affected?

Cheers

For people interested in it see https://youtu.be/km9CnGYBafI

**Comment by mookid8000** [comment]
I haven't tried it yet 🙂 but if I understand things correctly, it's not reflection as such that causes things to break, it's more when reflection is used to load types that the AOT analyzer cannot predict are going to be loaded... and I can't recall any situations where Rebus does that.

E.g. I would assume that the AOT system will know that your handler is needed, because you did something like this:

```csharp
services.AddRebusHandler<YourMessageHandler>();
```

so that type and its dependencies should be kept. But I might be wrong 😅 it would be fun to try though.

**Comment by dguisinger** [comment]
@SbiCA Did you ever have a chance to try out Rebus with AOT?  Was there a determination of whether or not it works?

**Comment by SbiCA** [comment]
@dguisinger  nope I'm afraid not so if you made first experience I'd also be interested 😅

**Comment by dguisinger** [comment]
@SbiCA hah, I've recently ripped out AutoMapper and MediatR because they didn't work with AoT and the project owner doesn't want to put time into making it work.   I then spent time looking into Mass Transit and then realized it also doesn't work with AoT and their project owner has a similar attitude.   I've pretty much given up on non-Microsoft frameworks at this point and don't want to waste more time evaluating things just to find they don't work with modern framework runtime options.  Kind of disappointed MSFT just said they are holding back their event framework due to complaints... I bet it would have worked just fine out of the box.

**Comment by SbiCA** [comment]
@dguisinger  agreed that's a pity because IHMO I think it could become a benefit for libs as well so you don't have to maintain all that and it could be more focused on the thing the do best.  I like rebus for it's "simplicity" and the docs for instance.

Not sure what @mookid8000  thinks of the development?
To me that looks like an missed opportunity as well. see https://github.com/dotnet/aspnetcore/issues/53219#issuecomment-1921960530 

I think it could be some advantage but then on the I'm not sure how much this would push rebus towards something more integrated such as https://wolverinefx.net/ 🤔

**Comment by mookid8000** [comment]
@SbiCA I would love to support AOT, but I haven't tried it, so I don't even know if it maybe already works?

Has anyone tried publishing with the flag turned on? if so, how did it go?

**Comment by SbiCA** [comment]
@mookid8000  sorry for the async feedback loop didn't try it so far because I know in our project we first have to ged rid of a few other dependencies that will for sure not work nicely with AOT. 
I'll try to get a minimal and isolated sample on version on the weekend since so far no one else seems to have tried it. 

as a side quest did you already checkout https://github.com/dotnet/aspnetcore/issues/53219#issuecomment-1921960530  the new "messaging" middleware and if it's going to happen because if so that might save a lot of work 🤔

**Comment by mookid8000** [comment]
@SbiCA Yeah I followed that thread since the beginning 😅 

It's of course interesting to see how Microsoft's common eventing framework is going to turn out. From @davidfowl's code sketch it sure looks like it could be super neat and easy to get started with. For now, I'm just sitting here with 🍿 , and then I'll see what opportunities unfold as it progresses.

**Comment by SbiCA** [comment]
@mookid8000  update from my end: sorry didn't get to it due to other prios(interrupts) I'll try again the following two weeks 🤞

**Comment by SbiCA** [comment]
@mookid8000  finally got some time.

here is a working sample with in memory transport but I had to use STJ 

https://github.com/SbiCA/Rebus-NativeAot/blob/74c9d37479f3256809040434a62c3eab91021620/Program.cs#L13

I'd still like to give it a try with a real world transport to make sure those work as well, but I guess at least a good start 😄

**Comment by SbiCA** [comment]
@mookid8000  just finished sqs play time and it seems to work as well (at least with this minimal demo) 

https://github.com/SbiCA/Rebus-NativeAot/blob/f90610ef8ec00dd42fb79b171975d75661b456a0/Program.cs#L30   

Not sure if we can say it's safe to use with native AOT or can you see some troublesome scenarios e.g. TypeBasedRouter using reflection that wouldn't be allowed?
